frontend/src/bundles/studio/components/timeline/timecursor/timecursor.tsx

```diff
@@ -1,46 +1,83 @@
+    useEffect,
     useLayoutEffect,
     useRef as useReference,
     useState,
     useTimelineContext,
 } from '~/bundles/common/hooks/hooks.js';
+import { MILLISECONDS_PER_REFRESH } from '~/bundles/studio/constants/constants.js';
+import { actions as studioActions } from '~/bundles/studio/store/studio.js';
 import styles from '~/framework/theme/styles/css-modules/timeline.module.css';
 
 type Properties = {
     interval?: number;
 };
 
 const TimeCursor: React.FC<Properties> = ({ interval }) => {
+    const dispatch = useAppDispatch();
+    const { isPlaying, elapsedTime, duration } = useAppSelector(
+        ({ studio }) => ({
+            isPlaying: studio.player.isPlaying,
+            elapsedTime: studio.player.elapsedTime,
+            duration: studio.player.duration,
+        }),
+    );
+
     const timeCursorReference = useReference<HTMLDivElement>(null);
-    const renderTimeReference = useReference(Date.now());
+    const renderTimeReference = useReference(0);
     const { range, direction, sidebarWidth, valueToPixels, pixelsToValue } =
         useTimelineContext();
 
     const side = direction === 'rtl' ? 'right' : 'left';
-    const millisecondPerRefresh = 1000;
+
     const [isDragging, setIsDragging] = useState(false);
     const [cursorPosition, setCursorPosition] = useState<number | null>(null);
 
+    useEffect(() => {
+        if (elapsedTime >= duration) {
+            void dispatch(studioActions.setPlaying(false));
+        }
+    }, [dispatch, elapsedTime, duration]);
+
     useLayoutEffect(() => {
         const offsetCursor = (): void => {
             if (!timeCursorReference.current || cursorPosition !== null) {
                 return;
             }
-            const timeDelta = Date.now() - renderTimeReference.current;
+
+            const currentTime = Date.now();
+            const timeDelta =
+                currentTime - renderTimeReference.current + elapsedTime;
             const timeDeltaInPixels = valueToPixels(timeDelta);
 
             const sideDelta = sidebarWidth + timeDeltaInPixels;
             timeCursorReference.current.style[side] = `${sideDelta}px`;
+
+            dispatch(studioActions.setElapsedTime(timeDelta));
+            renderTimeReference.current = currentTime;
         };
-        offsetCursor();
-        const cursorUpdateInterval = setInterval(
-            offsetCursor,
-            interval ?? millisecondPerRefresh,
-        );
+
+        let cursorUpdateInterval: NodeJS.Timeout | null = null;
+
+        if (isPlaying) {
+            renderTimeReference.current = Date.now();
+
+            offsetCursor();
+
+            cursorUpdateInterval = setInterval(
+                offsetCursor,
+                interval ?? MILLISECONDS_PER_REFRESH,
+            );
+        }
+
         return () => {
-            clearInterval(cursorUpdateInterval);
+            if (cursorUpdateInterval) {
+                clearInterval(cursorUpdateInterval);
+            }
         };
     }, [
```

Comment:
Maybe here you could use requestAnimationFrame and create a custom hook. 
https://layonez.medium.com/performant-animations-with-requestanimationframe-and-react-hooks-99a32c5c9fbf 

Also, you can check my PR with the dnd component. I used this hook there, and you can modify it and use it for your component as well. 
https://github.com/BinaryStudioAcademy/bsa-2024-outreachvids/pull/98/files#diff-6c59c5f2fd31b508b5b28947fd4aa51601ae0c874128ddf782694391489d5972